### PR TITLE
Correcting kmeans documentation 

### PR DESCRIPTION
**Pull request by StrahinjaLukic**
Correcting the documentation of `cluster.vq.kmeans` as per issue #2279 .

### CODE REVIEW

**Review by perimosocordiae — APPROVED** [review]
Looks right to me, thanks!

**Comment by mdhaber** [comment]
In "Background", "defined as the sum of the squared distances between each observation vector and its dominating centroid" still appears. Shouldn't this be changed, too?

**Comment by StrahinjaLukic** [comment]
Thanks for your input, you are right about the minimization objective function. It is the sum of the squared distances, not the average distance.

My primary observation was that the second return from kmeans, called "distortion", is neither the "sum of the squared differences", nor the mean squared distance, nor even the RMS. It is the mean plain old non-squared distance. This is demonstrated with the trivial 1D test with 1 cluster shown below.

However, in line with your comment and the documentation, the same test shows that the minimization objective function is the sum of the squared distances. 

The example:

    >>> set = array([[0., ],
    ...              [0., ],
    ...              [3., ]])
    >>> 
    >>> codes, d = kmeans(set, 1)
    >>> codes
    array([[1.]])
    >>> d
    1.3333333333333333
    >>> vq(set, codes)
    (array([0, 0, 0], dtype=int32), array([1., 1., 2.]))

Note that in the terminology of the vq module, "code" means the cluster center. This comes from the information theory jargon, as explained in "Background". 

So the center of the single cluster in this example is 1. This is unsurprising, as 1 is the arithmetic mean of the set of observations (0,0,3). This also shows that the sum of the squared distances is used as the minimization objective. If the average distance was used as the minimization objective, the cluster center would be at zero. So far so good.

Now notice that the second return, called "distortion", is 1.3333333333333333. This is nothing else but the mean distance from the observations to the cluster center. The distances (non-squared) are given in the second return from vq(set, codes) in the example. The sum of the squared distances is 6, the mean squared distance is 2, the RMS distance is sqrt(2), while the mean non-squared distance is 1.333...

I observed this consistence of the second return from kmeans with the mean non-squared distance on several real cases in 3D. So I am reasonably confident that this is indeed the definition of this return value.

But I put my correction in the wrong place. In the context of the sentence that I initially corrected, the "distortion" is used to mean the minimization objective function.

I will now push a new commit with my updated proposed corrections. Let me know what you think.

PS. You have probably noticed by now that I use the term "distance" for its plain meaning in the Euclidean space, i.e. the (non-squared) magnitude of the difference of two vectors. The "distance" array returned by the vq function also contains non-squared distances. The corrections in my PR are consistent with this definition.

PPS.  Speaking of the "sum of the squared **differences**": Note that **differences** between points in n>1 dimensions are not defined. One might talk about differences of individual coordinates, but that would unnecessarily complicate the explanation. I prefer to say "sum of the squared distances" for the minimization objective function.

**Comment by mdhaber** [comment]
@StrahinjaLukic You said "I observed this consistence of the second return from kmeans with the mean non-squared distance on several real cases in 3D. So I am reasonably confident that this is indeed the definition of this return value."
Thanks! The code seems to agree (I trace it below). 

But is this what you would _want_ the second output of `kmeans` to be?
Or would it make more sense for it to be the actual objective function?

---

The function `kmeans` says that its second output is `distortion` (line 375), which is really called `best_dist` inside the function (464). `best_dist` is the best value of `dist` (463), the second output of `_kmeans`, after `_kmeans` is run `iter` times.

The docustring of `_kmeans` calls the second output `avg_dist` (279) and defines it as
```
    avg_dist
        the average distance a observation is from a code in the book.
        Lower means the code_book matches the data better.
```
So this would be consistent with your findings. As the docstrings are already known to be inconsistent, we'd better look further : ) `avg_dist` is actually `prev_avg_dists[1]`, which comes from line 311:
```
prev_avg_dists.append(distort.mean(axis=-1))
```
So lit looks like it is indeed a mean of something!

`distort` is the second output of `vq` (310), which calls its second output `dist` (186):
```
    dist : ndarray
        The distortion (distance) between the observation and its nearest
        code.
```
Depending on the data type, this either comes from `_vq` (Line 211, imported from another file) or `py_vq` (212), which calls it (on line 239):
```
    mind_dist : ndarray
        min_dist[i] gives the distance between the ith observation and its
        corresponding code.
```
`mind_dist` comes from taking a minimum over `dist` (264), which comes from taking `scipy.spatial.distance.cdist` between the `obs` and `code_book`. 
So I believe your change is correct, at least if `py_vq` is called and if there are no mistakes in the code.

**Comment by StrahinjaLukic** [comment]
@mdhaber , thanks for looking all the way up in the code!

What would I want the second output of kmeans to be? It seems to me that the most useful output (besides the list of centroids) is actually the list of "labels" telling which cluster which observation belongs to. This is, for example, returned by kmeans2. The list of labels can also be calculated by the user, but at the cost of repeating the calculation of distances (squared or non-squared). This consumes CPU time and leads to the creation of excess code. 

It is interesting for some use cases to have kmeans return also the value of the objective function. This was proposed in issue #2279.
I am not sure whether the "average distance" that is returned at present is useful at all. 

In any case, I would not change the minimization objective function.

**Comment by mdhaber** [comment]
I agree.

Regarding returning the value of the objective function: why modify the documentation when we could actually fix the bug? This is what the public documentation says is supposed to happen, so I think it's safe to fix that.

As for returning the list of labels, `kmeans2` already does that, so I wonder why that was not part of the original plan for `kmeans`. I understand that returning a new third parameter could break existing code, but perhaps an optional argument like `return_labels` (default `False`) could be set `True` to get those.

What do you think?
In any case, thanks for your patience and help!

**Comment by StrahinjaLukic** [comment]
I agree on all points. I will now close this PR. Let me know if I can assist.

**Comment by mdhaber** [comment]
@StrahinjaLukic Certainly you're welcome to assist : ) I mean, if you're willing to make the change, I'd be happy to review again. I'd suggest fixing the returned "distortion" in one PR and returning the list of labels in another.

**Comment by StrahinjaLukic** [comment]
Ok. Will do so.

**Comment by StrahinjaLukic** [comment]
Ok, looking through the code with the goal to modify it so that it returns the "minimization objective function" value I realized the following: Kmeans does not explicitly minimize any function. It performs the optimization using an algorithm that iterates the following two steps:

1. Distances of observations from the centroids are calculated and each observation is assigned a centroid it belongs to. For this, `vq.vq()` is called. We may call this step the _membership update_.
2. Each centroid is updated to the mean of the observations currently belonging to the centroid. This is done by calling `_vq.update_cluster_means()`. We may call this step the _centroid update_.

1 and 2 are repeated until the (non-squared) average distance is (almost) not changing anymore, i.e. it is stable. Note that the iteration stops when the average distance is stable, but in general this does not happen at a local minimum value of the average distance! The criterion is stability, not minimum. This is the answer why in my trivial example the resulting centroid is not 0 but 1. With the "centroid" at 0, the average distance would be at its minimum. With the "centroid" at 1, it is not, but it is stable over the iterations outlined above.

On the other hand, when the average distance is stable, the sum of the squared distances is at a local minimum, despite the fact that it is not even calculated at any point of the algorithm. This is a simple consequence of two facts:

1. At convergence, the cluster "membership" of the observation points is stable.  An exception are observation points at nearly the same distance from two or more centroids. Membership update for such points has little influence on the average distance or on the squared distances. 
2. Since the centroid update sets the centroids to the Euclidean mean location of the corresponding cluster, the sum of the squared distances **within each cluster** is at a local minimum at the end of every iteration of the algorithm. At convergence, the sum of the squared distances of all clusters is also at a local minimum.

A remaining question for this discussion is the following: If we want to return the sum of the squared distances from `kmeans()`, we need to add an extra calculation step after convergence. I am not sure if it is necessary to add this calculation regardless of the needs of the specific application. If the user needs the sum of the squared distances, they can use `vq.vq()` to get the array of distances and calculate the sum of the squared distances from it. So with this reasoning, I am inclined to reopen my suggestion to correct the documentation. This time around, I think that the brief description of the kmeans algorithm should also be updated. What do you think?

**Comment by mdhaber** [comment]
OK. I agree now that updating the documentation only is appropriate. A reference would be good for "On the other hand, when the average distance is stable, the sum of the squared distances is at a local minimum." but in any case I think it's an improvement.

**Comment by mdhaber** [comment]
@StrahinjaLukic The changes look good to me. I agree now with the decision to change only the documentation. The changes to the documentation fix the original issue and also improve it otherwise. I think you've changed the documentation in all the places needed.
@perimosocordiae I see you reviewed this before, so I thought I'd reach out as I'm new to the core team. - should I just click the big green Merge button? : )

File: scipy/cluster/vq.py [hunk 1]
```diff
@@ -322,14 +321,15 @@ def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True):
...
 
-    The k-means algorithm adjusts the centroids until sufficient
-    progress cannot be made, i.e. the change in distortion since
-    the last iteration is less than some threshold. This yields
+    The k-means algorithm adjusts the classification of the observations
+    into clusters and updates the cluster centroids until the position of
+    the centroids is stable over successive iterations. In this 
+    implementation of the algorithm, the stability of the centroids is 
+    determined by comparing the absolute value of the change in the average 
+    (non-squared) distance between the observations and their corresponding
```
Comment (perimosocordiae) [inline scipy/cluster/vq.py]: How about "average Euclidean distance" here?

File: scipy/cluster/vq.py [hunk 2]
```diff
@@ -373,8 +373,11 @@ def kmeans(obs, k_or_guess, iter=20, thresh=1e-5, check_finite=True):
        not necessarily the globally minimal distortion.
 
     distortion : float
-       The distortion between the observations passed and the
-       centroids generated.
+       The mean (non-squared) Euclidean distance
+       between the observations passed and the
```
Comment (perimosocordiae) [inline scipy/cluster/vq.py]: super-nit: these lines could be wrapped more nicely

**Comment by perimosocordiae** [comment]
@mdhaber Thanks for pinging me on this. I left some very minor comments, but overall I'm +1 to merge.

I typically go with the "Squash and merge" option, but if the PR author has been careful to keep a tidy commit history with meaningful messages I'll use the "Rebase and merge". I don't think we've standardized on an approach.

**Comment by StrahinjaLukic** [comment]
I made the changes that @perimosocordiae suggested. 
Some of my commit messages refer to sentences I added and then deleted as I learned details about the algorithm along the way. Perhaps those messages are not very useful for the history...

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
@StrahinjaLukic That's fine, the "Squash and merge" option makes it easy for us to merge a single coherent commit, regardless of the state of the actual commits in this PR.

**Comment by ilayn** [comment]
Thanks all!

**Pull request merged**